backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/objcTypes.kt

```diff
@@ -51,16 +52,45 @@ class ObjCClassType(
     }
 }
 
-class ObjCGenericTypeDeclaration(
-        val typeParameterDescriptor: TypeParameterDescriptor,
-        val namer: ObjCExportNamer
-) : ObjCNonNullReferenceType() {
+sealed class ObjCGenericType : ObjCNonNullReferenceType() {
+    abstract val variance: Variance
+}
+
+data class ObjCGenericRawType(
```

Comment:
Generics of class declarations are not types, they are type parameter declarations. Mixing declarations into types representation class hierarchy seems wrong to me. Please create separate class representing Obj-C type parameter.

Note: generics of super class at `@interface` declaration are still types. More formally, they are type arguments, but in absence of use-site type variance it is ok to say they are types.

Reply:
Good point. I refactored the generic declarations into a separate hierarchy that ensures generic declarations don't mix with types:
```
Renderable
+-- ObjCType
|   +-- ObjCRawType
|   +-- ObjCReferenceType
|   |   +-- ObjCNonNullReferenceType
|   |   |   +-- ObjCClassType
|   |   |   +-- ObjCGenericTypeUsage
|   |   |   |   +-- ObjCGenericTypeRawUsage
|   |   |   |   +-- ObjCGenericTypeParameterUsage
|   |   |   +-- ObjCProtocolType
|   |   |   +-- ObjCIdType
|   |   |   +-- ObjCInstanceType
|   |   |   +-- ObjCBlockPointerType
|   |   |   +-- ObjCMetaClassType
|   |   +-- ObjCNullableReferenceType
|   +-- ObjCPrimitiveType
|   +-- ObjCPointerType
|   +-- ObjCVoidType
+-- ObjCGenericTypeDeclaration
    +-- ObjCGenericTypeRawDeclaration
    +-- ObjCGenericTypeParameterDeclaration
```

Does that address your concerns?

Reply:
Mostly: https://github.com/JetBrains/kotlin-native/pull/4727#discussion_r586393731

Reply:
Updated hierarchy:

```
ObjCType
+-- ObjCRawType
+-- ObjCReferenceType
|   +-- ObjCNonNullReferenceType
|   |   +-- ObjCClassType
|   |   +-- ObjCGenericTypeUsage
|   |   |   +-- ObjCGenericTypeRawUsage
|   |   |   +-- ObjCGenericTypeParameterUsage
|   |   +-- ObjCProtocolType
|   |   +-- ObjCIdType
|   |   +-- ObjCInstanceType
|   |   +-- ObjCBlockPointerType
|   |   +-- ObjCMetaClassType
|   +-- ObjCNullableReferenceType
+-- ObjCPrimitiveType
+-- ObjCPointerType
+-- ObjCVoidType

ObjCGenericTypeDeclaration
+-- ObjCGenericTypeRawDeclaration
+-- ObjCGenericTypeParameterDeclaration
```